From 7d1703d5460fa46b742fadcc5b6388953980ac67 Mon Sep 17 00:00:00 2001
From: Viktoria Maximova <viktoria.maksimova@intel.com>
Date: Wed, 4 Oct 2023 16:25:38 +0200
Subject: [PATCH 157/172] Reverse translation of access + load/store operations
 for cooperative matrix (#2165)

Implement translation via SPIR-V friendly calls, as:

the LLVM instructions are not capable to accept target extension types;
cooperative matrix is an opaque object and accessing elements is implementation defined, hence we can't use GEP to which AccessChain naturally maps, since GEP has a different meaning.
As for now some BE would need to recognize and define what to do with a call to __spirv_AccessChain(matrix, index). Better option is to map such SPIR-V to an intrinsic or define an appropriate type in LLVM (hence defining rules for GEP and other instructions) , but it's off the table now.
---
 lib/SPIRV/SPIRVReader.cpp                              |  4 ++++
 .../SPV_KHR_cooperative_matrix/access_store.ll         | 10 +++++++---
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/lib/SPIRV/SPIRVReader.cpp b/lib/SPIRV/SPIRVReader.cpp
index a0693ba0..79afc6c9 100644
--- a/lib/SPIRV/SPIRVReader.cpp
+++ b/lib/SPIRV/SPIRVReader.cpp
@@ -2112,6 +2112,10 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
     auto AC = static_cast<SPIRVAccessChainBase *>(BV);
     auto Base = transValue(AC->getBase(), F, BB);
     SPIRVType *BaseSPVTy = AC->getBase()->getType();
+    if (BaseSPVTy->isTypePointer() &&
+        BaseSPVTy->getPointerElementType()->isTypeCooperativeMatrixKHR()) {
+      return mapValue(BV, transSPIRVBuiltinFromInst(AC, BB));
+    }
     Type *BaseTy =
         BaseSPVTy->isTypeVector()
             ? transType(
diff --git a/test/transcoding/SPV_KHR_cooperative_matrix/access_store.ll b/test/transcoding/SPV_KHR_cooperative_matrix/access_store.ll
index 551a17ac..4f17cac1 100644
--- a/test/transcoding/SPV_KHR_cooperative_matrix/access_store.ll
+++ b/test/transcoding/SPV_KHR_cooperative_matrix/access_store.ll
@@ -3,9 +3,8 @@
 ; RUN: llvm-spirv %t.spv -to-text -o %t.spt
 ; RUN: FileCheck < %t.spt %s --check-prefix=CHECK-SPIRV
 
-; TODO: come up with an approach and implement reverse translation
-; R/UN: llvm-spirv -r %t.spv -o %t.rev.bc
-; R/UN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefix=CHECK-LLVM
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefix=CHECK-LLVM
 
 ; CHECK-SPIRV: TypeInt [[#TypeInt:]] 32 0
 ; CHECK-SPIRV-DAG: Constant [[#TypeInt]] [[#Const0:]] 0
@@ -24,6 +23,11 @@
 ; CHECK-SPIRV: AccessChain [[#TypeIntPtr]] [[#Res:]] [[#VarMatrixPtr]] [[#Const1]]
 ; CHECK-SPIRV: Store [[#Res]] [[#Const42]]
 
+; CHECK-LLVM: %0 = alloca %spirv.CooperativeMatrixKHR._int_3_12_12_0 addrspace(1)*
+; CHECK-LLVM: %Obj = call spir_func %spirv.CooperativeMatrixKHR._int_3_12_12_0 addrspace(1)* @_Z26__spirv_CompositeConstructi(i32 0)
+; CHECK-LLVM: store %spirv.CooperativeMatrixKHR._int_3_12_12_0 addrspace(1)* %Obj, %spirv.CooperativeMatrixKHR._int_3_12_12_0 addrspace(1)** %0
+; CHECK-LLVM: %call = call spir_func i32* @_Z19__spirv_AccessChainPPU3AS143__spirv_CooperativeMatrixKHR__int_3_12_12_0i(%spirv.CooperativeMatrixKHR._int_3_12_12_0 addrspace(1)** %0, i32 1)
+; CHECK-LLVM: store i32 42, i32* %call
 
 target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
 target triple = "spir64-unknown-unknown"
-- 
2.20.1

